gitreceived/gitreceived.go

```diff
@@ -307,14 +313,134 @@ func ensureCacheRepo(path string) error {
 	defer cacheMtx.Unlock()
 
 	cachePath := *repoPath + "/" + path
-	if _, err := os.Stat(cachePath); os.IsNotExist(err) {
-		os.MkdirAll(cachePath, 0755)
+
+	// remove old copy
+	if _, err := os.Stat(cachePath); err == nil {
+		os.RemoveAll(cachePath)
+	}
+	os.MkdirAll(cachePath, 0755)
+
+	res, err := http.Get("http://blobstore.discoverd/cache/" + path + ".tar")
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == 200 {
+		// cache hit
+		r := tar.NewReader(res.Body)
+		for {
+			header, err := r.Next()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+
+			filename := cachePath + "/" + header.Name
+
+			switch header.Typeflag {
+			case tar.TypeDir:
+				if _, err := os.Stat(filename); os.IsNotExist(err) {
+					if err := os.MkdirAll(filename, os.FileMode(header.Mode)); err != nil {
+						return err
+					}
+				}
+				if err := os.Chmod(filename, os.FileMode(header.Mode)); err != nil {
```

Comment:
Chmod should be there, in case we created the directory under ``case tar.TypeReg`, so that we now set it to the correct permissions.